engine/execution/ingestion/stop/stop_control.go

```diff
@@ -0,0 +1,660 @@
+	stopAfterExecuting flow.Identifier
+
+	// if the stop parameters were set by the version beacon
+	fromVersionBeacon bool
+}
+
+// String returns string in the format "crash@20023[versionBeacon]" or
+// "stop@20023@blockID[manual]"
+// block ID is only present if stopAfterExecuting is set
+// the ID is from the block that should be executed last and has height one
+// less than StopBeforeHeight
+func (s stopBoundary) String() string {
+	if !s.Set() {
+		return "none"
+	}
+
+	sb := strings.Builder{}
+	if s.ShouldCrash {
+		sb.WriteString("crash")
+	} else {
+		sb.WriteString("stop")
+	}
+	sb.WriteString("@")
+	sb.WriteString(fmt.Sprintf("%d", s.StopBeforeHeight))
+
+	if s.stopAfterExecuting != flow.ZeroID {
+		sb.WriteString("@")
+		sb.WriteString(s.stopAfterExecuting.String())
+	}
+	if s.fromVersionBeacon {
+		sb.WriteString("[versionBeacon]")
+	} else {
+		sb.WriteString("[manual]")
+	}
+
+	return sb.String()
+}
+
+// StopControlHeaders is an interface for fetching headers
+// Its jut a small subset of storage.Headers for comments see storage.Headers
+type StopControlHeaders interface {
+	ByHeight(height uint64) (*flow.Header, error)
+}
+
+// NewStopControl creates new StopControl.
+//
+// We currently have no strong guarantee that the node version is a valid semver.
+// See build.SemverV2 for more details. That is why nil is a valid input for node version
+// without a node version, the stop control can still be used for manual stopping.
+func NewStopControl(
+	log zerolog.Logger,
+	exeState state.ReadOnlyExecutionState,
+	headers StopControlHeaders,
+	versionBeacons storage.VersionBeacons,
+	nodeVersion *semver.Version,
+	latestFinalizedBlock *flow.Header,
+	withStoppedExecution bool,
+	crashOnVersionBoundaryReached bool,
+) *StopControl {
+	// We should not miss block finalized events, and we should be able to handle them
+	// faster than they are produced anyway.
+	blockFinalizedChan := make(chan *flow.Header, 1000)
+
+	sc := &StopControl{
+		log: log.With().
+			Str("component", "stop_control").
+			Logger(),
+
+		blockFinalizedChan: blockFinalizedChan,
+
+		exeState:                      exeState,
+		headers:                       headers,
+		nodeVersion:                   nodeVersion,
+		versionBeacons:                versionBeacons,
+		stopped:                       withStoppedExecution,
+		crashOnVersionBoundaryReached: crashOnVersionBoundaryReached,
+		// the default is to never stop
+		stopBoundary: stopBoundary{
+			StopParameters: StopParameters{
+				StopBeforeHeight: NoStopHeight,
+			},
+		},
+	}
+
+	if sc.nodeVersion != nil {
+		log = log.With().
+			Stringer("node_version", sc.nodeVersion).
+			Bool("crash_on_version_boundary_reached",
+				sc.crashOnVersionBoundaryReached).
+			Logger()
+	}
+
+	log.Info().Msgf("Created")
+
+	cm := component.NewComponentManagerBuilder()
+	cm.AddWorker(sc.processEvents)
+	cm.AddWorker(func(ctx irrecoverable.SignalerContext, ready component.ReadyFunc) {
```

Comment:
Is there a reason to do it this way? Usually we place initialization code straight in the constructor.

Reply:
the reason is that it can fail, which is why I wanted to use `ctx irrecoverable.SignalerContext`